### PR TITLE
Expand default ruff checks

### PR DESCRIPTION
**Pull request by amitschang**
We seem to be missing out on a lot of good checks and hints by using only the defaults. This expands the set while setting up some ignores.

However, don't fail the check except for the ones we are used to.

**Notes**:
* The tox check is meant to be more compatible with what we already have, so it wont fail but just report new violations. In this way it is more for hints in the IDE and setting some shared guidelines (e.g. in what we choose to check/ignore)
* For whole new projects, we might want to remove the `--select` from tox check-style and be a bit stricter from the start, but that can be per project.
* For existing projects we can backport this and keep the tox check, which should open up some nice hints, and any desired fixes can be done along the way.
* For things already transitioned, we wouldn't bother backporting.
* ❗ I just chose a handful of ignores that I don't care too much about, happy for updates to that set 😄

### CODE REVIEW

**Comment by jamienoss** [comment]
@amitschang See #129

**Comment by jamienoss** [comment]
Re https://github.com/ssec-jhu/base-template/pull/129#issuecomment-2037357329

>BTW, in the pre-push, could we even add the --fix option. I guess by this time the commit is there and could generate a diff. Wouldn't want to be too heavy handed though...

I think this should be manual to avoid compounding failing situations when stuff's being pushed that shouldn't be etc, rebases that fail, stuff like that.

File: pyproject.toml
Comment (jamienoss) [inline pyproject.toml:102]: I'm really not a fan of this (unless I've misunderstood it). I've stuck with the C styling where strings get double and chars get single.

Comment (jamienoss) [reply pyproject.toml:102]: I'm also just not a fan of single quotes period lol

**Review by jamienoss — CHANGES_REQUESTED** [review]
Thanks again for adding these! 👍 

A few comments about the ignores, though my major concern is the usage of running the linter twice, at least for the git hook (which I know I added, but we can discuss here).

File: githooks/pre-push
Comment (jamienoss) [inline githooks/pre-push:8]: @amitschang FYI I requested changes based on my own code lol See https://github.com/ssec-jhu/base-template/pull/128#discussion_r1551735773

I think running it twice will be confusing when stuff fails but doesn't block.

File: pyproject.toml
Comment (jamienoss) [reply pyproject.toml:102]: Also, 🙏 please don't make me have to change 😭

**Comment by amitschang** [comment]
And BTW, this is not meant to be final. We absolutely should keep modifying the ignore list when ruff adds new stuff and when we come across a rule we don't care too much about. I'd rather ignore than have to sprinkle `#noqa` all over the place 😱 !

File: pyproject.toml
Comment (jamienoss) [inline pyproject.toml:102]: BTW, I don't want to make other people use double quotes if they're not used to or don't want to - so I was thinking more along the lines of turning this off.

File: tox.ini
Comment (jamienoss) [inline tox.ini:15]: Also... sorry not to have stated this initially, I expected some discussion around this, but isn't the use of only E and F here misleading since the ``pyproject.toml`` config selects all?

File: pyproject.toml
Comment (amitschang) [reply pyproject.toml:102]: sure, we can remove the check altogether if ppl don't want to converge. But I do feel like specifically these format-only kind of rules, like whitespace, naming conventions, quotes, line-spilling style, etc. we could agree to be group-consistent (in our own stuff).

File: tox.ini
Comment (amitschang) [reply tox.ini:15]: this is primarily to transition. Then we can backport to existing projects while letting them still build without intervention. This repo even has a number of violations with new rules.

I was aiming to treat these as useful hints that show up in IDE and that we iterate toward. For example there is a set of pandas rules that seem pretty cool, but I don't know about a hard enforcement on those yet, especially in existing repos.

I would encourage any new work we do to eliminate the selections (as stated in comment), and if by so doing we find annoying ones rules pop up, we come back here and update the ignores.

**Review by ryanhausen — COMMENTED** [review]
I like the idea of being more lenient with style on tests. 

However, I think there should be some discussion on style in general. If it's the case that we want style to be something we enforce, we should pick a consistent general attitude. For example, if there is more than one way to do something, we either take the stance "any option is fine" or "please use A over B..G".  

I think when it comes to some decisions about style my preference would be to have as many formatting rules as possible enforced by a tool. `ruff` has a formatter which by default is black style, but there are other tools too. Disagreements about style aren't always super productive and letting a tool handle it and "own" the rules is an easy way to handle that.

https://docs.astral.sh/ruff/formatter/#philosophy
https://github.com/google/yapf
https://black.readthedocs.io/en/stable/index.html
https://blue.readthedocs.io/en/latest/

File: pyproject.toml
Comment (amitschang) [reply pyproject.toml:59]: I'm mostly OK with type-hints, but we would need to pick and choose these rules, for instance: https://docs.astral.sh/ruff/rules/missing-type-self/ would drive me bananas and I won't stand for it 😱 . https://docs.astral.sh/ruff/rules/missing-type-kwargs/ also seems strange to me.. If we pick the basics, lets go for it 😄

**Comment by amitschang** [comment]
@ryanhausen, I think I agree with you - pure-formatting things we should be able agree on some specifics and stick with those, all other things we don't bother about. We can perhaps limit that to the set of "fixable" things. We already have some of this codified in "E" and "F" rules, I suspect we should have some more over time (e.g. some more flake8 and docs rules).

The "ALL" here contains things that are also not pure-formatting, like pandas/numpy/alogorithm complexity etc, and that is where I would be less strict - so think of them as hints. On the other hand, perhaps that could be bothersome, and I am not a fan of requiring sprinkling noqa comments all over.

But for this, other than the annotations thing, did you have any concrete change you'd like to see in this?

File: pyproject.toml
Comment (ryanhausen) [reply pyproject.toml:59]: Agreed. I think the rebels amongst us can also use `Any`.  self and kwargs do seem a bit much haha

**Review by ryanhausen — APPROVED** [review]
Thanks Arik!

**Comment by ryanhausen** [comment]
I don't want to hold this up. This seems like it will be a longer conversation. Happy to discuss more.

**Review by jamienoss — APPROVED** [review]

**Comment by amitschang** [comment]
@ryanhausen, I know you once approved - I've since added fewer annotation exclusions. There's no rush to get this in so feel free to say your piece now. Of course, this will have basically no immediate effect and you can always make another PR if that'll work better.

**Review by ryanhausen — APPROVED** [review]
@amitschang Thanks for checking in!

**Pull request merged**
